### PR TITLE
record build durations properly on build completion

### PR DESCRIPTION
**Pull request by bparees**
fixes https://github.com/openshift/origin/issues/12549

### CODE REVIEW

**Comment by bparees** [comment]
@csrwng ptal
@derekwaynecarr looks like you added the original Decorator implementation which i'm removing because it doesn't do the right thing (seems like what it does is update the duration value as soon as we update the build the first time and then never again, which means the build durations get reported as very short values), so would like your sign off also in case i'm missing something critical about why it was done in that way.

**Comment by bparees** [comment]
Note that I also updated the describer logic to not use the duration field, in anticipation of deprecating it in the future.

File: pkg/build/controller/controller.go [hunk 1]
```diff
@@ -75,6 +75,9 @@ func (bc *BuildController) CancelBuild(build *buildapi.Build) error {
 	build.Status.Phase = buildapi.BuildPhaseCancelled
 	now := unversioned.Now()
 	build.Status.CompletionTimestamp = &now
+	if build.Status.StartTimestamp != nil {
+		build.Status.Duration = build.Status.CompletionTimestamp.Rfc3339Copy().Time.Sub(build.Status.StartTimestamp.Rfc3339Copy().Time)
```
Comment (csrwng) [inline pkg/build/controller/controller.go]: seems like a good candidate for a utility function

**Comment by bparees** [comment]
> As long as we set the start/stop time, the duration should get calculated on retrieval.

yeah now that i understand where it's being invoked (and that it's not storing the result), i don't get why it's not working..(and to be fair, I didn't dig deeply into it, so i'd be curious if you can reproduce what I was seeing) but regardless it seems like this PR is a more straightforward implementation.

**Comment by csrwng** [comment]
@bparees so the issue is that the decorator is only calculating the duration when the completionTime is empty. However, if you always set it, then you get the right result. See my debug change here: https://github.com/csrwng/origin/commit/668567b0b722f9f2c4b186a56aa2d2efec4302da

**Comment by bparees** [comment]
@csrwng the weird thing is I was seeing a value like "7s" (when the build was complete, so completion timestamp was set, but the delta between start time and completion time was way more than 7s).  But again, I didn't have a chance to really debug into it.


I still prefer the pattern of updating the duration explicitly since we're already updating other status information about the build, so while the decorator approach has some wins in that you don't have to ensure you set duration in every codepath where a build might complete, it also means you're operating on the build in a totally different codepath from anything else.

(Hopefully we can just kill "duration" in the future....)

**Comment by csrwng** [comment]
@bparees so it looks like we were never storing the duration (maybe at some point we were?). The decorator was only setting the duration on retrieval if the completion time was not set. At some point during the build it looks like the build object was retrieved, the decorator populated it, and then it was stored.

If you factor out the utility func to calculate the duration, this pull LGTM

**Comment by bparees** [comment]
[testextended][extended:core(s2i build with a quota)]

**Comment by derekwaynecarr** [comment]
@bparees -- i have literally no memory of doing anything in this space, so will let others sign off.

**Comment by bparees** [comment]
[merge]

**Comment by bparees** [comment]
flake https://github.com/openshift/origin/issues/12558
flake https://github.com/openshift/origin/issues/9414

**Comment by bparees** [comment]
flake https://github.com/openshift/origin/issues/12558
flake https://github.com/openshift/origin/issues/11887
[merge]

**Comment by bparees** [comment]
flake https://github.com/openshift/origin/issues/12558
[merge]

**Pull request merged**